lib/link_accounts_page.dart

```diff
@@ -1,134 +1,194 @@
+import 'dart:async';
+
 import 'package:firebase_auth/firebase_auth.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter_auth_buttons/flutter_auth_buttons.dart';
 import 'package:flutter_facebook_login/flutter_facebook_login.dart';
 import 'package:google_sign_in/google_sign_in.dart';
 
-class LinkAccountsPage extends StatelessWidget {
-  static final routeName = '/link_accounts';
-
-  @override
-  Widget build(BuildContext context) {
-    return Scaffold(
-      appBar: AppBar(),
-      body: FutureBuilder(
-          future: FirebaseAuth.instance.currentUser(),
-          builder: (context, snapshot) {
-            if (!snapshot.hasData) {
-              return CircularProgressIndicator();
-            } else {
-              final user = snapshot.data as FirebaseUser;
-              return Material(
-                child: Column(
-                  mainAxisAlignment: MainAxisAlignment.center,
-                  children: <Widget>[
-                    if (!_hasLinkedProvider('google.com', user.providerData))
-                      GoogleSignInButton(
-                        onPressed: () async {
-                          _linkGoogle(context, user).listen((event) {
-                            print(event);
-                          });
-                        },
-                      ),
-                    if (!_hasLinkedProvider('facebook.com', user.providerData))
-                      FacebookSignInButton(
-                        onPressed: () {
-                          _linkFacebook(context, user).listen((event) {
-                            print(event);
-                          });
-                        },
-                      ),
-                  ],
-                ),
-              );
-            }
-          }),
-    );
-  }
+enum UIStatus {
+  loading,
+  done,
+  error,
 }
 
-bool _hasLinkedProvider(String id, List<UserInfo> providersInfo) {
-  for (final info in providersInfo) {
-    if (info.providerId == id) {
-      return true;
-    }
-  }
-  return false;
-}
-
-Stream<int> _linkGoogle(BuildContext context, FirebaseUser user) async* {
-  try {
-    final _googleSignIn = GoogleSignIn(scopes: <String>['email']);
-    final _googleUser = await _googleSignIn.signIn();
-
-    // if the user canceled signin, an error is thrown but it gets swallowed
-    // by the signIn() method so we need to reset the UI and close the stream
-    if (_googleUser == null) {
-      yield 0;
-      return;
-    }
+class LinkAccountsPage extends StatelessWidget {
+  static final routeName = '/link_accounts';
+  final StreamController<UIStatus> _uiController = StreamController();
```

Comment:
Is not closing the controller intentional? I can't see it causing any problems in this case but personally I think it's still good practice to always close controllers. I think this is a good reference (he was on the Dart team at the time): https://stackoverflow.com/a/49884064/1992736

Reply:
I definitely forgot about closing the StreamController. It should be fixed now.